Mathlib/RingTheory/PowerSeries/Restricted.lean

```diff
@@ -0,0 +1,135 @@
+Authors: William Coram
+-/
+
+import Mathlib.Analysis.Normed.Group.Ultra
+import Mathlib.Analysis.RCLike.Basic
+import Mathlib.RingTheory.PowerSeries.Basic
+import Mathlib.Tactic.Bound
+
+/-!
+# Restricted power series
+
+`IsRestricted` : We say a power series over a normed ring `R` is restricted for a parameter `c` if
+`‖coeff R i f‖ * c ^ i → 0`.
+
+-/
+
+namespace PowerSeries
+
+variable {R : Type*} [NormedRing R] (c : ℝ)
+
+open PowerSeries Filter
+open scoped Topology
+
+/-- A power series over `R` is restricted of paramerter `c` if we have
+`‖coeff R i f‖ * c ^ i → 0`. -/
+def IsRestricted (f : PowerSeries R) :=
+  Tendsto (fun (i : ℕ) => (norm (coeff R i f)) * c ^ i) atTop (𝓝 0)
+
+namespace IsRestricted
+
+lemma isRestricted_iff {f : PowerSeries R} : IsRestricted c f ↔
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → ‖‖(coeff R n) f‖ * c ^ n‖ < ε := by
+  simp [IsRestricted, NormedAddCommGroup.tendsto_atTop]
+
+lemma isRestricted_iff_abs (f : PowerSeries R) : IsRestricted c f ↔ IsRestricted |c| f := by
+  simp [isRestricted_iff]
+
+lemma zero : IsRestricted c (0 : PowerSeries R) := by
+  simp [IsRestricted]
+
+lemma one : IsRestricted c (1 : PowerSeries R) := by
+  simp only [isRestricted_iff, coeff_one, norm_mul, norm_pow, Real.norm_eq_abs]
+  refine fun _ _ ↦ ⟨1, fun n hn ↦ ?_ ⟩
+  split
+  · omega
+  · simpa
+
+lemma add {f g : PowerSeries R} (hf : IsRestricted c f) (hg : IsRestricted c g) :
+    IsRestricted c (f + g) := by
+  simp only [isRestricted_iff, map_add, norm_mul, norm_pow, Real.norm_eq_abs] at ⊢ hf hg
+  intro ε hε
+  obtain ⟨fN, hfN⟩ := hf (ε / 2) (by positivity)
+  obtain ⟨gN, hgN⟩ := hg (ε / 2) (by positivity)
+  simp only [abs_norm] at hfN hgN ⊢
+  refine ⟨max fN gN, fun n hn ↦ ?_ ⟩
+  calc _ ≤ ‖(coeff R n) f‖ * |c| ^ n + ‖(coeff R n) g‖ * |c| ^ n := by grw [norm_add_le, add_mul]
+       _ < ε / 2 + ε / 2 := by gcongr <;> grind
+       _ = ε := by ring
+
+lemma neg {f : PowerSeries R} (hf : IsRestricted c f) : IsRestricted c (-f) := by
+  simpa [isRestricted_iff] using hf
+
+
+/-- The set of `‖coeff R i f‖ * c ^ i` for a given power series `f` and parameter `c`. -/
+def convergenceSet (f : PowerSeries R) : Set ℝ := {‖coeff R i f‖ * c^i | i : ℕ}
+
+open Finset in
+lemma convergenceSet_BddAbove {f : PowerSeries R} (hf : IsRestricted c f) :
+  BddAbove (convergenceSet c f) := by
+  simp_rw [isRestricted_iff] at hf
+  obtain ⟨N, hf⟩ := by simpa using (hf 1)
+  rw [bddAbove_def, convergenceSet]
+  use max 1 (max' (image (fun i ↦ ‖coeff R i f‖ * c ^ i) (range (N + 1))) (by simp))
+  simp only [Set.mem_setOf_eq, le_sup_iff, forall_exists_index, forall_apply_eq_imp_iff]
+  intro i
+  rcases le_total i N with h | h
+  · right
+    apply le_max'
+    simp only [mem_image, mem_range]
+    exact ⟨i, Order.lt_add_one_iff.mpr h, rfl⟩
+  · left
+    calc _ ≤ ‖(coeff R i) f‖ * |c ^ i| := by bound
+         _ ≤ 1 := by simpa using (hf i h).le
+
+lemma convergenceSet_leNNeg {f : PowerSeries R} (hf : IsRestricted c f) :
```

Comment:
I really think this lemma should be removed: you prove that a certain set `S` is bounded, and then you explicitly say that there is a positive `A` such that for all `s ∈ S`, `s ≤ A`. This is of course totally unrelated to your set `S`, so if you really need it it should be state in full generality. But in my opinion you should just use the `x` you constructed here in the lemma `mul` below.